<a href="https://colab.research.google.com/github/bzznrc/Python-101/blob/master/FetchDemo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fetch Demo 🐕

In [2]:
from google.colab import files
uploaded = files.upload()
#print(list(uploaded)[0]) #Gets the key

Saving FetchUS_BAR_SKUs.csv to FetchUS_BAR_SKUs.csv


In [3]:
import pandas as pd
import numpy as np
import io

#Magical bit of code you shouldn't really worry about
df = pd.read_csv(io.StringIO(uploaded['FetchUS_BAR_SKUs.csv'].decode('utf-8')))
#Ta-daaa! This displays the first few lines of your dataframe
df.head()

barcode type category_1            category_2    category_3  \
0  16202004180  UPC       FOOD  PASTA SHAPE AND TYPE     SPAGHETTI   
1  33617000026  UPC       FOOD               CRACKER  CRACKERBREAD   
2  33617000453  UPC       FOOD               CRACKER        SESAME   
3  33617000460  UPC       FOOD               CRACKER  CRACKERBREAD   
4  33617000484  UPC       FOOD               CRACKER  CRACKERBREAD   

  manufacturer    brand  item_count    size product_name  \
0      BARILLA  BARILLA         1.0   12 OZ      BARILLA   
1      BARILLA     WASA         1.0  9.2 OZ      BARILLA   
2      BARILLA     WASA         1.0  6.7 OZ      BARILLA   
3      BARILLA     WASA         1.0  6.7 OZ      BARILLA   
4      BARILLA     WASA         1.0  6.1 OZ      BARILLA   

                                      description  
0            BARILLA PASTA VEGGIE SPAGHETTI 12 OZ  
1                       WASA WHOLE GRAIN - 9.2 OZ  
2           WASA SESAME THINS 190G X 10  - 6.7 OZ  
3  WASA ROSEM & SALT THINS 190G X 10 USA - 6.7 OZ  
4           WASA SESAME GLUTEN FREE 175G X 10 USA

# Wait. What? 🤷‍♂️
Oook. Bit of background needed.

We're now working with some real-world data from one of our real-world partners. The reason we're importing this data with a very old fashioned .csv is because these data are quite sensitive, so we couldn't really dump them into the GitHub.

The bits of code up there just let you upload a file from you own PC and load its content into a DataFrame.

### But What is a DataFrame? 🖼
Right. So you can see there were a couple of libraries imported there. Pandas is the most important out of them, and it's by far the #1 used library for all the Data Scientists in the world.
The main thing it does, it's introducing this DataFrame concept into Python, and all the functionalities to work with it.

A DataFrame is, basically, a data table. And you can think of it as an "Excel within Python", if you'd like. As you can see in the output above, the column names are the ones coming from your .csv while the rows are indexed with a number (the one on the left).

Our DataFrame is called "df", as pretty much every other DataFrame in every notebook in the world. We're data scientists, so you can guess creativity is not top-of-mind.

In [4]:
#When you have a DataFrame the first thing you want to do is see what's inside.
#The head() method shows you the first 5 lines
df.head()

barcode type category_1            category_2    category_3  \
0  16202004180  UPC       FOOD  PASTA SHAPE AND TYPE     SPAGHETTI   
1  33617000026  UPC       FOOD               CRACKER  CRACKERBREAD   
2  33617000453  UPC       FOOD               CRACKER        SESAME   
3  33617000460  UPC       FOOD               CRACKER  CRACKERBREAD   
4  33617000484  UPC       FOOD               CRACKER  CRACKERBREAD   

  manufacturer    brand  item_count    size product_name  \
0      BARILLA  BARILLA         1.0   12 OZ      BARILLA   
1      BARILLA     WASA         1.0  9.2 OZ      BARILLA   
2      BARILLA     WASA         1.0  6.7 OZ      BARILLA   
3      BARILLA     WASA         1.0  6.7 OZ      BARILLA   
4      BARILLA     WASA         1.0  6.1 OZ      BARILLA   

                                      description  
0            BARILLA PASTA VEGGIE SPAGHETTI 12 OZ  
1                       WASA WHOLE GRAIN - 9.2 OZ  
2           WASA SESAME THINS 190G X 10  - 6.7 OZ  
3  WASA ROSEM & SALT THINS 190G X 10 USA - 6.7 OZ  
4           WASA SESAME GLUTEN FREE 175G X 10 USA

In [5]:
#Unimpressingly, the tail() method shows you the last 5 lines
df.tail()

barcode type category_1            category_2      category_3  \
6876  706010115887  UPC       FOOD  PASTA SHAPE AND TYPE         LASAGNE   
6877  724923453407  UPC       FOOD               CRACKER    CRACKERBREAD   
6878  855691001177  UPC       FOOD    PASTA PIZZA SAUCES  MARINARA SAUCE   
6879  863669743509  UPC       FOOD               CRACKER    CRACKERBREAD   
6880  891128133011  UPC       FOOD         PASTA DINNERS     READY PASTA   

     manufacturer    brand  item_count                   size product_name  \
6876      BARILLA  BARILLA         1.0                   1 LB      BARILLA   
6877      BARILLA     WASA         1.0                 5.4 OZ         WASA   
6878      BARILLA  BARILLA         1.0                  24 OZ      BARILLA   
6879      BARILLA     WASA         1.0  9.1999999999999993 OZ         WASA   
6880      BARILLA  BARILLA         1.0                 8.5 OZ      BARILLA   

                                            description  
6876                    BARILLA PASTA WAVY LASAGNE 1 LB  
6877            WASA ORIGINAL GLUTEN FREE 155G X 10 USA  
6878  BARILLA PASTA AND PIZZA SAUCE PREMIUM MARINARA...  
6879                          WASA WHOLE GRAIN - 9.2 OZ  
6880                 BARILLA Gemelli Ready Pasta 8.5 oz

In [6]:
#The describe method provide you with a summary of how the amounts are distributed in the numerical columns
#This is very useful to quickly assess the range of values, what the averages are, how narrow the distributions might be
df.describe()

barcode   item_count
count  6.881000e+03  6853.000000
mean   2.092765e+11     1.717934
std    5.340253e+11     1.802726
min    1.111061e+10     1.000000
25%    7.680800e+10     1.000000
50%    7.680801e+10     1.000000
75%    7.680852e+10     1.000000
max    6.674100e+12    10.000000

In [7]:
#Lastly, the shape attribute. Be aware, this is not a method, so there are no parenthesis required
#An attribute is basically a label that your object (your df in this case) always carries
df.shape

#The result is a tuple with (rows, columns). If you need any of those actual values on their own:
#df.shape[0] #Gets you the nr. of rows
#df.shape[1] #Gets you the nr. of columns

(6881, 11)

#Selecting Stuff 👉
So you have a bunch of data into a DataFrame. What your inner Data Analyst will likely want to do is either selecting and displaying data that match a certain criteria and then plotting the results.

Let's start with selecting stuff. Let's say we want only the Cracker products.

This kinda looks like what we did with if conditions, but here when you write a condition, it is applied to every row. This is how you ask Pandas whether each line is a Cracker or not:

In [8]:
df['category_2'] == 'CRACKER'

0       False
1        True
2        True
3        True
4        True
        ...  
6876    False
6877     True
6878    False
6879     True
6880    False
Name: category_2, Length: 6881, dtype: bool

Cool! So now we have 6881 True or False values and we don't have a clue what to do with them!

Kidding, now we'll store the information that we got in a Series (which is, a DataFrame with a single column) and use it to select actual Cracker rows in the main DataFrame.

Basically what happens is that the DataFrame will only show rows that correspond to the "True" values in the Series:

In [9]:
#Same code as before, getting the True/False for our condition for all rows
isCracker = df['category_2'] == 'CRACKER'
#Using the results to display only the relevant rows
df[isCracker]

barcode type category_1 category_2    category_3 manufacturer  \
1       33617000026  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   
2       33617000453  UPC       FOOD    CRACKER        SESAME      BARILLA   
3       33617000460  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   
4       33617000484  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   
5       33617000491  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   
...             ...  ...        ...        ...           ...          ...   
6841  1030862270982  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   
6866   642526681822  UPC       FOOD    CRACKER        SESAME      BARILLA   
6870   688267159145  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   
6877   724923453407  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   
6879   863669743509  UPC       FOOD    CRACKER  CRACKERBREAD      BARILLA   

     brand  item_count                   size product_name  \
1     WASA         1.0                 9.2 OZ      BARILLA   
2     WASA         1.0                 6.7 OZ      BARILLA   
3     WASA         1.0                 6.7 OZ      BARILLA   
4     WASA         1.0                 6.1 OZ      BARILLA   
5     WASA         1.0                 5.4 OZ      BARILLA   
...    ...         ...                    ...          ...   
6841  WASA         1.0                 8.6 OZ      BARILLA   
6866  WASA         1.0                 6.7 OZ         WASA   
6870  WASA         1.0  9.6999999999999993 OZ         WASA   
6877  WASA         1.0                 5.4 OZ         WASA   
6879  WASA         1.0  9.1999999999999993 OZ         WASA   

                                         description  
1                          WASA WHOLE GRAIN - 9.2 OZ  
2              WASA SESAME THINS 190G X 10  - 6.7 OZ  
3     WASA ROSEM & SALT THINS 190G X 10 USA - 6.7 OZ  
4              WASA SESAME GLUTEN FREE 175G X 10 USA  
5            WASA ORIGINAL GLUTEN FREE 155G X 10 USA  
...                                              ...  
6841                          WASA THIN RYE - 8.6 OZ  
6866           WASA SESAME THINS 190G X 10  - 6.7 OZ  
6870                      WASA- Multigrain - 9.70 OZ  
6877         WASA ORIGINAL GLUTEN FREE 155G X 10 USA  
6879                       WASA WHOLE GRAIN - 9.2 OZ  

[236 rows x 11 columns]

# Grouping Stuff 🥂
How nice it is when our data come exactly in the format we need them into?
Not sure, never happened to me. 🤷‍♂️

Most likely, when you have a DataFrame you'll want to produce a few additional ones with different kinds of grouping, in order to see suff from different perspectives.

For example, with starting from the original DataFram we could ask:
- How many entries do we have for each category_3?
- How many different descriptions do we have for each category_3?

In order to answer to both questions, we'll definitely need to **group** the data by *category_3*.

Grouping the data by a field (or more than one fields) means producing a new DataFrame in which each row will correspond to only one grouping values.

Let's see an example:

In [22]:
df.groupby(by='category_3')['description'].count()

category_3
ALFREDO AND CREAM BASED SAUCE      60
ANGEL HAIR AND NOODLE STYLE       180
CRACKERBREAD                      207
ELBOWS AND MACARONI               360
FARFALLE                          210
FETTUCCINE                         90
FLAVORED WITH CHEESE               30
FLAVORED WITH MEAT                 30
FLAVORED WITH MUSHROOM             30
ITALIAN TRADITIONAL SAUCE         150
LASAGNE                           270
LINGUINE                          270
MARINARA SAUCE                    150
NONGRAIN PASTAS                   265
OTHER SAUCES                      180
OTHER STYLES                     1140
PASTA DINNERS                      90
PENNE                             630
PESTO                             180
READY PASTA                       112
RIGATONI                          150
ROTINI                            364
SESAME                             23
SHAPES AND ALPHABET                60
SHELL SHAPED                      210
SPAGHETTI                        1046
S

In [20]:
df.groupby(by='category_3')['description'].nunique()

category_3
ALFREDO AND CREAM BASED SAUCE     3
ANGEL HAIR AND NOODLE STYLE       8
CRACKERBREAD                      9
ELBOWS AND MACARONI              15
FARFALLE                          7
FETTUCCINE                        5
FLAVORED WITH CHEESE              1
FLAVORED WITH MEAT                1
FLAVORED WITH MUSHROOM            1
ITALIAN TRADITIONAL SAUCE         1
LASAGNE                          13
LINGUINE                         10
MARINARA SAUCE                    4
NONGRAIN PASTAS                  15
OTHER SAUCES                      6
OTHER STYLES                     40
PASTA DINNERS                     5
PENNE                            21
PESTO                             8
READY PASTA                       4
RIGATONI                          5
ROTINI                           15
SESAME                            1
SHAPES AND ALPHABET               4
SHELL SHAPED                     10
SPAGHETTI                        25
STICKS                            1
TORTELLINI AND RA

### Plotting 🖌🎨
Sooo tables are nice, but you know, other Data Analysts are doing all those crazy coloured graphs... 


In [11]:
import plotly.express as px

px.scatter(data_frame=df, x='category_1',y='item_count')

todo: 
- show how to get column (basket value)
- plot histogram on it
- group by store and count

- ex: what's the avg basket size for a certain supermarket?
- add a column, eg. percentage


- histograms
- add numerical feature

- if we had stores, we could have 2/3 numericals (barilla vol, total, other vol, total)